### DISCUSSION TITLE
pandoc vs djot document model / footnotes

### DISCUSSION THREAD

**Discussion by massifrg** [jgm/pandoc#9526] [Ideas]
I'm looking at the differences between pandoc and djot document models.

In this message I'm focusing on (foot)notes.

In pandoc, a Note is an Inline that contains a list of Blocks and has no attribute.
It's the only Inline element that can contain Blocks.

In djot, footnotes are split into a "Footnote reference" that has a "reference label" (a string) and a "Footnote" that contains a list a blocks and has the same reference label.

Djot's model opens the way to:

- **multiple references to the same footnote**; e.g. in ConTeXt there's `\footnote[label]{text}` for the first reference, and `\note[label]` for the next references to the same note;

- **multiple types of notes**, e.g. footnotes and end notes: by convention, you could give the same suffix to footnotes -- say "fn" -- and to end notes -- say "en" --, so that specific filters/writers could discriminate between the two kinds.

**Is pandoc going to cover djot's model?**

Something like this (in terms of Haskell constructor, as in `Text.Pandoc.Definition`):

`NoteRef label` (a new `Inline`)

`NoteText label [Block]` (a new `Block`) 

Or like this:

`Note label [Block]` (the current `Note` `Inline`, with a `label` string added)

`NoteRef label` (a new `Inline` for other references to the same note)

**Comment by jgm** [comment]
This would have been a more flexible choice to make at the beginning (and people have requested both of the things you mention -- see the issue tracker).  But this kind of change to the pandoc AST would be very painful to make, as it requires revision in virtually every pandoc module, plus many subsidary packages, as well as the whole ecosystem of filters &c that depend on pandoc.  Given the pain it would cause, and the effort it would take, I wouldn't count on this happening any time soon.

**Reply by massifrg** [reply]
Thank you for your quick reply.

In my editor of Pandoc AST, I'm simulating different kind of notes through this convention:

- if the `Note` content is a list of *only one* `Div` as its content, and that `Div` has a `note-type` attribute, then it can be treated as a different kind of note by filters/writers

- in all other cases, it's a plain footnote

It's a trick I learned from the `Div` with `custom-style` attribute around paragraphs that miss the `Attr` data structure.

That single `Div` as Note content should be transparent, though it's not _completely_: e.g. in Word footnotes get a sort of newline at the start of the footnote text, because of that surrounding block.
A filter could fix that in theory, at least for end notes, but unfortunately end notes are stored in an `endnotes.xml` file, not in the main `document.xml` file inside the `docx` package. It looks like filters can't access those inner files.
A filter can prepend a prefix to the footnotes content, and leave the differentiation task to a Word macro.

When the output format is HTML, you can do the job more easily with a filter.
Since I mostly typeset XML/XHTML with ConTeXt, I can use different kinds of notes that way.
